core/src/main/java/org/testcontainers/containers/ParsedDockerComposeFile.java

```diff
@@ -44,14 +54,73 @@
     }
 
     @VisibleForTesting
-    ParsedDockerComposeFile(Map<String, Object> testContent) {
-        this.composeFileContent = testContent;
+    ParsedDockerComposeFile(Map<String, Object> testContent, Map<String, String> env) {
+        //noinspection unchecked
+        this.composeFileContent = (Map<String, Object>) substituteEnv(testContent, env);
```

Comment:
Would it be possible to use Apache Commons [`StringSubstitutor`](https://commons.apache.org/proper/commons-text/apidocs/org/apache/commons/text/StringSubstitutor.html) instead? 

At a glance your implementation of `substituteEnv` looks decent, but it would be good to avoid having our own implementation of placeholder substitution from a maintenance/verification perspective.

Reply:
Using `StringSubtitutor` could be possible, but it seems to have a number of differences:

1. It behaves like `-` (replacing if the variable is set, even if empty). You can't really get the same behavior as `:-` as described in the docker docs.
2. Matching multiple different operators seem possible, albeit slightly weird.